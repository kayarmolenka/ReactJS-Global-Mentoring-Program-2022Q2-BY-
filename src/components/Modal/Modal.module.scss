@import "src/style/variables";
@import "src/style/mixins";

.modal {
  position: absolute;
  z-index: 4;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  background: $mainBlack;
  box-shadow: 0 0 2px rgba(0, 0, 0, 0.1), 0 2px 10px rgba(0, 0, 0, 0.07), 0 10px 20px rgba(0, 0, 0, 0.05), 0 10px 50px rgba(0, 0, 0, 0.05);
  padding: 50px;
  width: 754px;
  height: 840px;

  .modalCloseIcon {
    color: white;
    font-size: 37px;
    cursor: pointer;
  }

  .modalHeader {
    @include font-family;
    font-weight: 300;
    font-size: 40px;
    letter-spacing: 1px;
    text-transform: uppercase;
    color: $white;
    margin-bottom: 38px;
    display: flex;
    justify-content: space-between;
  }

  .modalInputsFields {
    display: grid;
    grid-template-columns: repeat(2, 1fr);
    grid-row-gap: 30px;
    grid-column-gap: 30px;
  }

  .modalContent {
    .modalContentLabel {
      @include font-family;
      letter-spacing: 0.888889px;
      text-transform: uppercase;
      color: $pink;
      mix-blend-mode: normal;
      opacity: 0.8;
      font-size: 16px;
      line-height: 20px;
      margin-bottom: 5px;
    }

    input {
      @include font-family;
      background: rgba(50, 50, 50, 0.948044);
      mix-blend-mode: normal;
      padding: 17px 0 16px 19px;
      border: none;
      border-radius: 4px;
      color: $white;
      font-size: 20px;
      line-height: 24px;
    }

    input::placeholder {
      @include font-family;
      font-size: 20px;
      line-height: 24px;
      color: $white;
      mix-blend-mode: normal;
      opacity: 0.3;
    }
  }

  .modalTitle {
    display: flex;
    flex-direction: column;

    input {
      padding: 17px 148px 16px 18px;
    }
  }

  .modalMovieUrl {
    display: flex;
    flex-direction: column;
  }

  .modalReleaseDate {
    display: flex;
    flex-direction: column;

    input {
      min-width: 301px;
    }

    input::-webkit-calendar-picker-indicator {
      color: $pink;
    }
  }

  .modalRating {
    display: flex;
    flex-direction: column;
  }

  .modalRuntime {
    display: flex;
    flex-direction: column;
  }

  .modalGenre {
    display: flex;
    flex-direction: column;
    height: 134px;
    gap: 6px;

    select {
      @include font-family;
      background: rgba(50, 50, 50, 0.948044);
      mix-blend-mode: normal;
      opacity: 0.8;
      border-radius: 4px;
      padding: 17px 0 16px 19px;
      color: $white;
      line-height: 24px;
      width: 100%;
    }

    .modalGenreIcon {
      color: $pink;
      font-size: 30px;
      position: relative;
      width: 50px;
      height: 30px;
      left: 347px;
      bottom: 47px;
      cursor: pointer;
      pointer-events: none;
    }
  }

  .modalGenreSelect {
    @include font-family;
    font-size: 16px;
    width: 196px;
    letter-spacing: 0.888889px;
    text-transform: uppercase;
    color: $white;
    border: none;
    cursor: pointer;
    appearance: none;
    background: rgba(35, 35, 35, 1) url("../../assets/images/arrow.png") no-repeat right;
  }

  .modalOverview {
    display: flex;
    flex-direction: column;

    textarea {
      background: rgba(50, 50, 50, 0.948044);
      mix-blend-mode: normal;
      border: none;
      border-radius: 4px;
      height: 177px;
      color: $white;
      font-size: 20px;
      line-height: 24px;
      padding: 10px;
    }

    .textareaErrorWarning {
      @include font-family;
      color: $pink;
      font-size: 14px;
      line-height: 17px;
      mix-blend-mode: normal;
      opacity: 0.8;
    }

    textarea::placeholder {
      @include font-family;
      font-size: 20px;
      line-height: 24px;
      color: $white;
      mix-blend-mode: normal;
      opacity: 0.3;
    }
  }

  .modalBtn {
    margin-top: 60px;
    display: flex;
    justify-content: end;
    gap: 13px;

    .modalBtnReset {
      @include font-family;
      padding: 17px 57px;
      font-size: 20px;
      line-height: 24px;
      text-transform: uppercase;
      color: $pink;
      border: 1px solid $pink;
      border-radius: 4px;
      background: $mainBlack;
      cursor: pointer;
    }

    .modalBtnSubmit {
      @include font-family;
      padding: 17px 47px;
      font-size: 20px;
      line-height: 24px;
      text-align: center;
      text-transform: uppercase;
      color: $white;
      border: none;
      border-radius: 4px;
      background: $pink;
      cursor: pointer;
    }
  }
}

.modalShadow {
  background-color: rgba(0, 0, 0, 0.7);
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 3;
  cursor: auto;
}
