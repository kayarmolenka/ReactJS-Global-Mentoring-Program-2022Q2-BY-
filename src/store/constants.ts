import { IApplicationState } from './types';

export const initialState: IApplicationState = {
  movieList: [],
};

export const FETCH_MOVIE_LIST = 'movies/fetchMovieList';
export const FETCH_SORTED_MOVIE_LIST = 'movies/fetchSortedMovieList';
export const FETCH_FILTERED_MOVIE_LIST = 'movies/fetchFilteredMovieList';
export const API_URL = 'http://localhost:4000/';
