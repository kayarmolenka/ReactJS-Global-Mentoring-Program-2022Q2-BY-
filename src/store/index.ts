export * from './store';
export * from './constants';
export * from './thunks';
export * from './types';
export * from './movieListSlice';
export * from './selectors';
