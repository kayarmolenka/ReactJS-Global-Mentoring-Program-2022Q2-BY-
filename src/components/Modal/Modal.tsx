import { ChangeEvent, useCallback, useState } from 'react';
import { faXmark } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Button, MultipleDropdown } from '../index';
import { ButtonType } from '../../models';
import { valueFilter } from '../../constants';

import styles from './Modal.module.scss';

interface IInitialStateMovieDescription {
  title: string;
  releaseDate: string;
  genre: string[];
  runtime: number | string;
  overview: string;
  rating: number | string;
  movieUrl: string;
}

interface IModalProps {
  textHeader: string;
  isOpenModal: boolean;
  setIsOpenModal: (data: boolean) => void;
  setSuccessModal: (data: boolean) => void;
  initialState?: IInitialStateMovieDescription;
}

export const Modal = (props: IModalProps) => {
  const { textHeader, isOpenModal, setIsOpenModal, setSuccessModal, initialState } = props;
  const [genreIsOpen, setGenreIsOpen] = useState(false);
  const [isShowValidationError, setIsShowValidationError] = useState(false);
  const initState: IInitialStateMovieDescription = initialState || {
    title: '',
    releaseDate: '',
    movieUrl: '',
    rating: '',
    genre: [],
    runtime: '',
    overview: '',
  };

  const [stateForm, setStateForm] = useState(initState);

  const genresValue = [
    ...new Set([
      ...valueFilter
        .filter((genre) => genre !== 'All')
        .concat(initialState?.genre || [])
        .sort(),
    ]),
  ];

  const closeModalWindow = useCallback(() => {
    setIsOpenModal(false);
    setGenreIsOpen(false);
    setStateForm(initState);
    setIsShowValidationError(false);
    document.body.style.overflow = 'auto';
  }, [setIsOpenModal]);

  const stopClosed = useCallback((e) => {
    e.stopPropagation();
  }, []);

  const onHandleFormItems = (
    e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement>,
  ) => {
    setStateForm((state) => ({
      ...state,
      [`${e.target.name}`]: e.target.value,
    }));
  };

  const handleCheckbox = (e: ChangeEvent<HTMLInputElement>) => {
    const clickedGenre = e.target.name;

    if (stateForm.genre.includes(clickedGenre)) {
      setStateForm((state) => ({
        ...state,
        genre: [...state.genre.filter((genre) => genre !== clickedGenre)],
      }));
    } else {
      setStateForm((state) => ({
        ...state,
        genre: [...state.genre, clickedGenre],
      }));
      setIsShowValidationError(false);
    }
  };

  const switchToDateType = (e: ChangeEvent<HTMLInputElement>) => {
    e.target.type = 'date';
  };

  const switchToTextType = (e: ChangeEvent<HTMLInputElement>) => {
    e.target.type = 'text';
    stateForm.releaseDate ? (e.target.value = String(stateForm.releaseDate)) : '';
  };

  const completeAddMovie = () => {
    setGenreIsOpen(false);
    if (!stateForm.genre.length) {
      setIsShowValidationError(true);
    } else {
      document.body.style.overflow = 'auto';
      setIsOpenModal(false);
      setSuccessModal(true);
      console.log(stateForm); //TODO for verify the data after submitting the form
      setStateForm(initState);
    }
  };

  return isOpenModal ? (
    <div className={styles.modalShadow} onClick={closeModalWindow}>
      <div className={styles.modal} onClick={stopClosed}>
        <header className={styles.modalHeader}>
          {textHeader}
          <FontAwesomeIcon
            icon={faXmark}
            className={styles.modalCloseIcon}
            onClick={closeModalWindow}
          />
        </header>
        <div className={styles.modalContent}>
          <div className={styles.modalInputsFields}>
            <div className={styles.modalTitle}>
              <label htmlFor="title" className={styles.modalContentLabel}>
                Title
              </label>
              <input
                name="title"
                type="text"
                placeholder="Movie title"
                value={stateForm.title}
                onChange={onHandleFormItems}
              />
            </div>
            <div className={styles.modalReleaseDate}>
              <label htmlFor="release-date" className={styles.modalContentLabel}>
                Release date
              </label>
              <input
                name="releaseDate"
                type="text"
                placeholder="Select Date"
                value={stateForm.releaseDate}
                onFocus={switchToDateType}
                onBlur={switchToTextType}
                onChange={onHandleFormItems}
              />
            </div>
            <div className={styles.modalMovieUrl}>
              <label htmlFor="movie-url" className={styles.modalContentLabel}>
                Movie url
              </label>
              <input
                name="movieUrl"
                type="text"
                placeholder="htpps://"
                value={stateForm.movieUrl}
                onChange={onHandleFormItems}
              />
            </div>
            <div className={styles.modalRating}>
              <label htmlFor="rating" className={styles.modalContentLabel}>
                Rating
              </label>
              <input
                name="rating"
                type="text"
                placeholder="7.8"
                value={stateForm.rating}
                onChange={onHandleFormItems}
              />
            </div>
            <div className={styles.modalGenre}>
              <MultipleDropdown
                title="Select Genre"
                isDropdownOpen={genreIsOpen}
                setDropdownOpen={setGenreIsOpen}
                handleValue={handleCheckbox}
                choseGenres={stateForm.genre}
                isShowValidationError={isShowValidationError}
                setIsShowValidationError={setIsShowValidationError}
                genresValue={genresValue}
              />
            </div>
            <div className={styles.modalRuntime}>
              <label htmlFor="runtime" className={styles.modalContentLabel}>
                Runtime
              </label>
              <input
                name="runtime"
                type="text"
                placeholder="minutes"
                value={stateForm.runtime}
                onChange={onHandleFormItems}
              />
            </div>
          </div>
          <div className={styles.modalOverview}>
            <label htmlFor="overview" className={styles.modalContentLabel}>
              Overview
            </label>
            <textarea
              name="overview"
              placeholder="Movie description"
              value={stateForm.overview}
              onChange={onHandleFormItems}
            />
          </div>
          <div className={styles.modalBtn}>
            <Button
              className={styles.modalBtnReset}
              text="Reset"
              onClick={closeModalWindow}
              type={ButtonType.SUBMIT}
            />
            <Button
              className={styles.modalBtnSubmit}
              text="Submit"
              onClick={completeAddMovie}
              type={ButtonType.SUBMIT}
            />
          </div>
        </div>
      </div>
    </div>
  ) : null;
};
