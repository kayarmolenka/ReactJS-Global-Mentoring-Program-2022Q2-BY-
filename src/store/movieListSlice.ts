import { createSlice } from '@reduxjs/toolkit';
import { changeFilter, fetchMovieList, setGenre } from './thunks';
import { initialState } from './constants';

export const applicationSlice = createSlice({
  name: 'movies',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(setGenre, (state, action) => {
      state.activeGenre = action.payload;
    });

    builder.addCase(changeFilter, (state, action) => {
      state.sortingType = action.payload;
    });

    builder.addCase(fetchMovieList.fulfilled, (state, action) => {
      state.movieList = action.payload.data;
    });
  },
});

export const actionApplicationReducer = applicationSlice.reducer;
