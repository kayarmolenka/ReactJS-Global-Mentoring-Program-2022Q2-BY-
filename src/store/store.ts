import { useDispatch } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { AppDispatch, RootState } from './index';
import { actionApplicationReducer } from './movieListSlice';

export const reducer = {
  application: actionApplicationReducer,
};

export const createStore = (preloadedState?: Partial<RootState>) =>
  configureStore({ reducer, preloadedState });

export const store = createStore();

export const useAppDispatch = () => useDispatch<AppDispatch>();
