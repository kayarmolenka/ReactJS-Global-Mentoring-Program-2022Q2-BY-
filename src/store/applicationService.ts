import { IFetchMovieListResponse } from './applications';
import { useFetch } from '../hooks/useFetch';

export const applicationService = (
  url: string,
  init?: RequestInit,
): Promise<IFetchMovieListResponse> => {
  return useFetch(url, init);
};
