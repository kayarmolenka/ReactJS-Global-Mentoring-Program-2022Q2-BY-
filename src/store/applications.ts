import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { RootState } from './index';
import { applicationService } from './applicationService';
import { IMovieList } from '../models';

type ISortingType = string;

interface IApplicationState {
  movieList: IMovieList[];
}

export interface IFetchMovieListResponse {
  limit: number;
  offset: number;
  totalAmount: number;
  data: IMovieList[];
}

const initialState: IApplicationState = {
  movieList: [],
};

const FETCH_MOVIE_LIST = 'application/fetchMovieList';
const FETCH_SORTED_MOVIE_LIST = 'application/fetchSortedMovieList';
const FETCH_FILTERED_MOVIE_LIST = 'application/fetchFilteredMovieList';
const API_URL = 'http://localhost:4000/';

export const fetchMovieList = createAsyncThunk<
  IFetchMovieListResponse | undefined,
  void,
  { state: RootState; rejectedValue: string }
>(FETCH_MOVIE_LIST, async (_, thunkApi) => {
  try {
    return await applicationService(`${API_URL}movies?offset=1&limit=12`);
  } catch (error) {
    thunkApi.rejectWithValue((error as Error).message);
  }
});

export const fetchSortedMovieList = createAsyncThunk<
  IFetchMovieListResponse | undefined,
  ISortingType,
  { state: RootState; rejectedValue: string }
>(FETCH_SORTED_MOVIE_LIST, async (payload, thunkApi) => {
  try {
    return await applicationService(
      `${API_URL}movies?sortBy=${payload}&sortOrder=asc&offset=1&limit=12`,
    );
  } catch (error) {
    thunkApi.rejectWithValue((error as Error).message);
  }
});

export const fetchFilteredMovieList = createAsyncThunk<
  IFetchMovieListResponse | undefined,
  ISortingType,
  { state: RootState; rejectedValue: string }
>(FETCH_FILTERED_MOVIE_LIST, async (filterName, thunkApi) => {
  try {
    return await applicationService(`${API_URL}movies?filter=${filterName}&offset=1&limit=12`);
  } catch (error) {
    thunkApi.rejectWithValue((error as Error).message);
  }
});

export const applicationSlice = createSlice({
  name: 'application',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(fetchMovieList.fulfilled, (state, action) => {
      state.movieList = action.payload?.data || [];
    });

    builder.addCase(fetchSortedMovieList.fulfilled, (state, action) => {
      state.movieList = action.payload?.data || [];
    });

    builder.addCase(fetchFilteredMovieList.fulfilled, (state, action) => {
      state.movieList = action.payload?.data || [];
    });
  },
});

export const actionApplicationReducer = applicationSlice.reducer;
