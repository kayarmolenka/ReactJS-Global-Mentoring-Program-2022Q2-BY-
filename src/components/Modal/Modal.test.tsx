import { Provider } from 'react-redux';
import { render, screen } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import { Modal } from './Modal';
import { createStore } from '../../store';

describe('Modal', () => {
  const setIsOpenModal = jest.fn();
  const setSuccessModal = jest.fn();

  it('should render Add movie when isOpenModal equal true', () => {
    const { asFragment } = render(
      <Provider store={createStore()}>
        <Modal
          textHeader="Add Movie"
          isOpenModal={true}
          setIsOpenModal={setIsOpenModal}
          setSuccessModal={setSuccessModal}
        />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText('Add Movie')).toBeInTheDocument();
  });

  it('should not render Add movie when isOpenModal equal true', () => {
    act(() => {
      render(
        <Provider store={createStore()}>
          <Modal
            textHeader="Add Movie"
            isOpenModal={false}
            setIsOpenModal={setIsOpenModal}
            setSuccessModal={setSuccessModal}
          />
        </Provider>,
      );
    });

    expect(screen.queryByText('Add Movie')).not.toBeInTheDocument();
  });
});
