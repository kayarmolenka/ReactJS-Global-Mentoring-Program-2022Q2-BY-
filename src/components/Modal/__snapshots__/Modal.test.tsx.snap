// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should render Add movie when isOpenModal equal true 1`] = `
<DocumentFragment>
  <div
    class="modalShadow"
  >
    <div
      class="modal"
    >
      <header
        class="modalHeader"
      >
        Add Movie
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-xmark modalCloseIcon"
          data-icon="xmark"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M310.6 361.4c12.5 12.5 12.5 32.75 0 45.25C304.4 412.9 296.2 416 288 416s-16.38-3.125-22.62-9.375L160 301.3L54.63 406.6C48.38 412.9 40.19 416 32 416S15.63 412.9 9.375 406.6c-12.5-12.5-12.5-32.75 0-45.25l105.4-105.4L9.375 150.6c-12.5-12.5-12.5-32.75 0-45.25s32.75-12.5 45.25 0L160 210.8l105.4-105.4c12.5-12.5 32.75-12.5 45.25 0s12.5 32.75 0 45.25l-105.4 105.4L310.6 361.4z"
            fill="currentColor"
          />
        </svg>
      </header>
      <form
        action="#"
      >
        <div
          class="modalContent"
        >
          <div
            class="modalInputsFields"
          >
            <div
              class="modalTitle"
            >
              <label
                class="labelFormField"
                for="title"
              >
                Title
              </label>
              <input
                name="title"
                placeholder="Movie title"
                type="text"
                value=""
              />
            </div>
            <div>
              <div
                class="modalReleaseDate"
              >
                <label
                  class="labelFormField"
                  for="release_date"
                >
                  Release date
                </label>
                <input
                  name="release_date"
                  placeholder="Select Date"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="modalMovieUrl"
            >
              <label
                class="labelFormField"
                for="poster_path"
              >
                Movie url
              </label>
              <input
                name="poster_path"
                placeholder="htpps://"
                type="text"
                value=""
              />
            </div>
            <div
              class="modalRating"
            >
              <label
                class="labelFormField"
                for="vote_average"
              >
                Rating
              </label>
              <input
                name="vote_average"
                placeholder="7.8"
                type="text"
                value=""
              />
            </div>
            <div
              class="modalGenre"
            >
              <label
                class="modalLabel"
              >
                Genre
              </label>
              <div>
                <button
                  class="title"
                  type="button"
                >
                  Select Genre
                </button>
                <div
                  class="arrowIcon"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-caret-down "
                    data-icon="caret-down"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 320 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M310.6 246.6l-127.1 128C176.4 380.9 168.2 384 160 384s-16.38-3.125-22.63-9.375l-127.1-128C.2244 237.5-2.516 223.7 2.438 211.8S19.07 192 32 192h255.1c12.94 0 24.62 7.781 29.58 19.75S319.8 237.5 310.6 246.6z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div
              class="modalRuntime"
            >
              <label
                class="labelFormField"
                for="runtime"
              >
                Runtime
              </label>
              <input
                name="runtime"
                placeholder="minutes"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="modalOverview"
          >
            <label
              class="modalContentLabel"
              for="overview"
            >
              Overview
            </label>
            <textarea
              name="overview"
              placeholder="Movie description"
            />
          </div>
        </div>
        <div
          class="modalBtn"
        >
          <button
            class="modalBtnReset"
            type="submit"
          >
            Reset
          </button>
          <button
            class="modalBtnSubmit"
            type="submit"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
